Create Smart Battery Module

RKBoss6 opened the pull request:
This provides a module an app can use for battery. It learns your usage patterns over time by averaging out drainage rates, and extrapolates that into a pretty accurate prediction.

RKBoss6 commented:
I made logging an option in settings, app is good to go

RKBoss6 commented:
@RelapsingCertainly Logging shouldn't change anything fundamental, you're right, it's just for debugging. Give it up to 2 days to see if the drainage rate is just fluctuating, and it's still learning from that. I would love to hear a status update on this later on :) 

PS: how did you install the app? If you copied the files into the ide and uploaded it, maybe something went awry there...

RelapsingCertainly commented:
> PS: how did you install the app? If you copied the files into the ide and uploaded it, maybe something went awry there...

I did just download the files, rename them according to the metadata and uploaded them. I'll try to use the app loader in case that makes a difference

RKBoss6 commented:
@RelapsingCertainly Glad to hear! Let me know how it goes after a few days :)

RelapsingCertainly commented:
I was able to use the module for a bit and it works pretty well. I think it took around 6 hours to get a decent estimate and was surprisingly accurate.

P.s. if u are interested, i have made a clockinfo for smartbatt that would probably work well if bundled with the widget or module.

RKBoss6 commented:
@RelapsingCertainly That would be amazing! If you want, you could make a separate PR for it, or you could drop the code here and I can add it with you as a contributor?

RelapsingCertainly commented:
Heres the link to a gist with the clock info code. Could use some polish tbh: https://gist.github.com/RelapsingCertainly/03f91f0dbba33d940aad263416c77ad5

RKBoss6 commented:
> Heres the link to a gist with the clock info code. Could use some polish tbh: https://gist.github.com/RelapsingCertainly/03f91f0dbba33d940aad263416c77ad5

That looks really good! Maybe I'll add another clockInfo with current drainage rate, and days left as well :)

RelapsingCertainly commented:
Turns out redrawing was messed up so i fixed it in the gist. It should now update the clockinfo text and icon once a minute.

RelapsingCertainly commented:
> Maybe I'll add another clockInfo with current drainage rate, and days left as well :)

Current drainage sounds good, but what do you mean "and days left", like another card with the days remaining? (the existing card already does that) or a card with drainage and days remaining together?

RKBoss6 commented:
@RelapsingCertainly Sorry, forgot that the whole point of the module was to show the days remaining! lol...

RKBoss6 commented:
> Turns out redrawing was messed up so i fixed it in the gist. It should now update the clockinfo text and icon once a minute.

Nice! I'll get to work adding this into the module :)

RelapsingCertainly commented:
> @RelapsingCertainly Sorry, forgot that the whole point of the module was to show the days remaining! lol...

Lmfao, its cool 😂 

the clockinfo basically acts as a replacement for the battery clock info. It shows percentage when charging, days left if time left is more than 24hours or, hours left if there is less than a days amount of charge. In the future, a click might show percentage for a bit (like the widget) or open a battery app.

RKBoss6 commented:
@RelapsingCertainly I added the clockInfo, and added a second one for average battery drainage. Try reinstalling the app from [my app loader](https://rkboss6.github.io/BangleApps/?id=smartbatt), and check it out!

RelapsingCertainly commented:
The clock info is broken again, idk if it my watch or just the clock info but is not updating either on charge or to show the new estimate. 

P.s. Power manager might be broken since the battery percentage stays the same charging or not until its reloaded. Im currently running an experiment to see if 0.12 is indeed broken, ill report back with my finding. 

P.p.s turns out this module is a hard thing to test since my previous typical drainage would be around 5% per day.

RelapsingCertainly commented:
Well im back and can confirm, powermanager stable percentage is broken. After disabling it, both clock infos are working as they should.

RelapsingCertainly commented:
Have been using it for a bit, it looks like the time remaining only goes down.

So if i had 10 days of battery left and charge it, even if i now have 14 days, it will stay on 10 days remaining for 4 days until it loses another day.

RKBoss6 commented:
Btw, thanks for all the testing! I would have never caught these issues on my own :)

RelapsingCertainly commented:
Oh damn, recorder is a very good battery tool lol, it instantly lowered my estimation to 3 hrs so once its drained a little, im gonna put it on charge again and then see if it increases over 3hrs. 

If that fails, ill reinstall and such. 

Btw when do u recharge ur watch? Like do u let it die or like what is ur minimum time left before charging?

RelapsingCertainly commented:
> Btw, thanks for all the testing! I would have never caught these issues on my own :)

Well... have to find a way to delay making my own bangle apps somehow :P

bobrippling reviewed (commented):
LGTM, two smaller suggestions :)

RelapsingCertainly commented on apps/smartbatt/clkinfo.js:
```diff
@@ -0,0 +1,88 @@
+
+    var daysLeft = hrsLeft / 24;
+    daysLeft = Math.round(daysLeft);
+    if (Bangle.isCharging()) {
```
If i read it correctly, if bangle is charging then it shows the battery percentage otherwise its showing time estimation. 

Are you suggesting to only run the days left calulation if its not charging? Or am i misunderstanding the review?

bobrippling replied:
Ah no - there is only one place where we call `getHrsFormatted`:

https://github.com/espruino/BangleApps/blob/6aa202795eb0fe5ae4de3b9536283f849fdabafa/apps/smartbatt/clkinfo.js#L48-L53

So inside `getHrsFormatted`, we know `Bangle.isCharging()` will always be `false`

bobrippling replied:
Sweet, LGTM

RKBoss6 commented:
>Btw when do u recharge ur watch? Like do u let it die or like what is ur minimum time left before charging?

I generally charge it around 5-10%, and try charging it no more than 80% to protect the battery life. I think around 5h left is when I charge it

bobrippling commented on apps/smartbatt/clkinfo.js:
```diff
@@ -0,0 +1,88 @@
+    var s=24,g=Graphics.createArrayBuffer(24,24,1,{msb:true});
+    g.fillRect(0,6,s-3,18).clearRect(2,8,s-5,16).fillRect(s-2,10,s,15).fillRect(3,9,3+batt*(s-9)/100,15);
+    g.transparent=0;
+    img=g.asImage("string");
```
I've pushed [a litte extra to drop the global var](https://github.com/espruino/BangleApps/pull/3944/commits/175e2acaa71ffe9ada6dc44ae8c0efde98f3701d), hope you don't mind. If you're happy with that then let's merge it in :)

bobrippling replied:
Great, thanks!

bobrippling commented:
> I generally charge it around 5-10%, and try charging it no more than 80% to protect the battery life. I think around 5h left is when I charge it

Same for me. And thanks for checking the PR @RelapsingCertainly!

RKBoss6 commented:
> Well... have to find a way to delay making my own bangle apps somehow :P
@RelapsingCertainly Maybe you could try making a companion app for this module :) I am running low on time now, but it would be a perfect starter app to work on...

RKBoss6 commented:
Hmm, probably shows things like average drainage, timeleft, time when it's predicted to die, those kinds of things.. Maybe a big battery monitor in the center...
Have fun with it though, you do what you think would be best! Is this still your first app?

RelapsingCertainly commented:
it will be my first non-clock app, since I've made a really simple layout clock (not on app store). Mostly to use builtin lazy rending but also for future porting to the banglejs 3 whenever that happens. I went down a battery rabbit hole lol.

RKBoss6 commented:
Haha nice! Have fun!

RelapsingCertainly commented:
> Nice! That looks really good - quick question, is the graph based off logging data, and if so, does it hide if logging is disabled?

The graph will uses either the average drainage or E.getPowerUsage() to predict the future battery percentage. I ideally want this to work as both a companion app but also a general battery app.

RelapsingCertainly commented:
I've reached a decision that I wanted an opinion on, how often should I update the values?
I'm thinking:
- Graph is drawn at launch but then not updated. 
- Battery, Drainage and Time Left is updated once per minute.
- Usage is drawing once per second with the backlight on and, once per minute with the backlight off.

Last update is when the last smart battery prediction is from.

RKBoss6 commented:
> > Nice! That looks really good - quick question, is the graph based off logging data, and if so, does it hide if logging is disabled?
> 
> The graph will uses either the average drainage or E.getPowerUsage() to predict the future battery percentage. I ideally want this to work as both a companion app but also a general battery app. edit: added gps and hrm <img alt="download" width="176" height="176" src="https://private-user-images.githubusercontent.com/173109465/486550337-c1c777ef-f5a2-4ebe-8fe6-615137558ae0.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTcyODkzNTMsIm5iZiI6MTc1NzI4OTA1MywicGF0aCI6Ii8xNzMxMDk0NjUvNDg2NTUwMzM3LWMxYzc3N2VmLWY1YTItNGViZS04ZmU2LTYxNTEzNzU1OGFlMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTA3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwN1QyMzUwNTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01OTIyNDEwODBmZWE4ZDc5MmUzZDcwNTUxYzE3ZmE2YjFiZjI2ZGYxMjRmODIyMGEzZGY1ZTEzNWMzMDBmYTc1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.DqdILBaY5UZ796QK-4VzIZGfvN0cGb3S_v8ktBpQN-Q">

Hmm, wouldn't it make more sense to display past battery statistics, by tapping into `heath`'s battery logging? If you want a general battery app, I'd say that's better, because you can already infer the drainage from the average drain rate...

As for the updates, that all seems fine! Maybe consider adding colors, as that will also look nice, maybe black if off, or cyan/blue if on...

RelapsingCertainly commented:
i couldn't find a way to make past battery work since either the range varies too much (either 1 2% movement per day or 10% 15% per hour) so I decided to make it a wearos style battery page. The button will be used to enable or disable gps or hrm if they are running at the moment.

<img width="176" height="176" alt="download (2)" src="https://github.com/user-attachments/assets/ae4cf8e5-b231-454a-92b6-ae6264b71bf9" />

Inspiration:
<img width="396" height="396" alt="image" src="https://github.com/user-attachments/assets/80a51fb9-3708-4d80-b939-44a8cff9154a" />

RKBoss6 commented:
Nice! Is it possible to keep the time last updated? That seems very useful to me...

RelapsingCertainly commented:
I gotta actually add value updates and clean up the code. Im guessing there will be a draft by the end of the day. 

What should i name it? Im thinking battery manager (batman lol) but it doesnt really manage the battery right now.

RelapsingCertainly commented:
#3992 
just made a draft pull if you are interested

RKBoss6 commented:
I also see you changed your name - on the contrubitors section of the readme, do you want me to change the name there as well from relapsingcertainly?

RKBoss6 commented:
Of course!

sonicityV commented:
Oh hey, sorry about the late reply. My semester finals were happening this and last week so got caught up with that.

The device status that you tested was non fuctional since i was trying to think of ways to compact the device status (both on/off and the colour) functions into one call. (I think i might just call the function multiple times to reduce adding too many varriables ). 

As for the settings, i decided to not implement it rn since it would need to run in the background, write to storage so it can remember device status setting. Honestly the scope was just creeping wider and i had to make a call. 

The main goals of this app is, to view smartbar stats and lightweight enough to be bundled with the module if thats something you decide to do.

RKBoss6 commented:
No worries, looks good! When you have the time, maybe just polish it up a bit, and it should be good to go for a PR! 
Good judgement call - I don't think I would always want the app updating in the background anyway, so that's a good choice.

Overall it looks good,  sorry about pestering lol, just wanted to see the status of this :)
